spring-boot-project/spring-boot/src/main/java/org/springframework/boot/diagnostics/analyzer/MissingWebServerFactoryBeanFailureAnalyzer.java

```diff
@@ -0,0 +1,88 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.diagnostics.analyzer;
+
+import org.springframework.boot.WebApplicationType;
+import org.springframework.boot.diagnostics.AbstractFailureAnalyzer;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.boot.diagnostics.FailureAnalyzer;
+import org.springframework.boot.web.server.MissingWebServerFactoryBeanException;
+import org.springframework.context.ApplicationContextException;
+import org.springframework.util.ClassUtils;
+
+/**
+ * A {@link FailureAnalyzer} that performs analysis of failures caused by an
+ * {@link MissingWebServerFactoryBeanException}.
+ *
+ * @author Guirong Hu
+ */
+class MissingWebServerFactoryBeanFailureAnalyzer extends AbstractFailureAnalyzer<ApplicationContextException> {
+
+	private static final String WEBMVC_INDICATOR_CLASS = "org.springframework.web.servlet.DispatcherServlet";
+
+	private static final String WEBFLUX_INDICATOR_CLASS = "org.springframework.web.reactive.DispatcherHandler";
+
+	private final ClassLoader classLoader;
+
+	/**
+	 * Create a new {@code MissingWebServerFactoryBeanFailureAnalyzer} instance.
+	 */
+	MissingWebServerFactoryBeanFailureAnalyzer() {
+		this(null);
+	}
+
+	/**
+	 * Create a new {@code MissingWebServerFactoryBeanFailureAnalyzer} instance with the
+	 * given class loader.
+	 * @param classLoader the class loader
+	 */
+	MissingWebServerFactoryBeanFailureAnalyzer(ClassLoader classLoader) {
+		this.classLoader = classLoader;
+	}
+
+	@Override
+	protected FailureAnalysis analyze(Throwable rootFailure, ApplicationContextException cause) {
+		Throwable rootCause = cause.getCause();
+		if (rootCause instanceof MissingWebServerFactoryBeanException) {
+			WebApplicationType webApplicationType = ((MissingWebServerFactoryBeanException) rootCause)
+					.getWebApplicationType();
+			if (webApplicationType != null && webApplicationType != WebApplicationType.NONE) {
+				String action = action(webApplicationType);
+				if (action != null) {
+					return new FailureAnalysis(String.format(
+							"Reason: The running web application is of type %s, but the dependent class is missing.",
+							webApplicationType.name().toLowerCase()), action, cause);
+				}
+			}
+		}
+		return null;
+	}
+
+	private String action(WebApplicationType webApplicationType) {
+		String action = "Add `%s` dependencies to your application.";
+		if (webApplicationType == WebApplicationType.SERVLET
+				&& !ClassUtils.isPresent(WEBMVC_INDICATOR_CLASS, this.classLoader)) {
+			return String.format(action, "spring-boot-starter-web");
+		}
+		if (webApplicationType == WebApplicationType.REACTIVE
+				&& !ClassUtils.isPresent(WEBFLUX_INDICATOR_CLASS, this.classLoader)) {
+			return String.format(action, "spring-boot-starter-webflux");
+		}
+		return null;
+	}
```

Comment:
Being specific here will be tricky (see my comments above about the indicator classes). I think it might be better to display a more general action that recommends checking the configuration of the `web-application-type` property and the application's dependencies for a supported web server.

Reply:
The same as above.

Reply:
I am currently using `MissingWebServerFactoryBeanException#webServerFactoryClass` to determine the web application type, I am not sure if this is simpler, looking forward to your reply:)

Reply:
Done.